yogstation/code/modules/power/nuclear_engine.dm

```diff
@@ -0,0 +1,442 @@
+	ReactorInoperable = TRUE
+	priority_announce("Nuclear reactor status: CRITICAL Meltdown imminent! Evacuate engineering section IMMEDIATELY", "Nuclear Reactor Monitoring Subsystem",'sound/ai/attention.ogg')
+	addtimer(CALLBACK(src, .proc/Meltdown), 410) //It's the final countdown
+	for(var/mob/M in GLOB.player_list)
+		if(M.z == z)
+			M << null //Don't want to assault their eardrums
+			M << MeltDownSound
+
+/obj/machinery/power/NuclearReactor/proc/Meltdown() //qdel(station)
+	explosion(get_turf(src),20,40,30, 100, ignorecap=TRUE)
+	var/datum/round_event_control/radiation_storm/RS = new()
+	RS.runEvent()
+	fuel = null
+	monitor = null
+	waste = null
+	cut_overlays()
+	icon_state = "reactor-broken"
+	for(var/turf/T in orange(src, 10))
+		new /obj/effect/nuclearwaste(T)
+
+
+/obj/machinery/power/NuclearReactor/proc/CheckState()
+	if(state & FUELHATCH_OPEN || state & WASTEHATCH_OPEN)
+		radiation_pulse(src, Heat, 40) //You fucked up kid
+	if(Heat >= MELTDOWN) //Oh FUCK
+		MeltdownAnnounce()
+		return
+	if(Heat >= CRITICAL_HEAT)
+		radiation_pulse(src, Heat*2, 60)
+		if(world.time >= SavedWarningTime + WarningSoundCooldown)
+			SavedWarningTime = world.time
+			for(var/mob/M in GLOB.player_list)
+				if(M.z == z)
+					M << WarningSound
+		else
+			return
+
+/obj/machinery/power/NuclearReactor/proc/ProcessAtmos()
+	for(var/datum/gas_mixture/S in outlet.airs)
+		if(S.temperature <= 20000) //A small nerf to avoid ambient temperatures in the room reaching 2000 degrees
+			var/num = Heat
+			if(num <= 0)
+				num = 0
+			S.temperature += num
+
+/obj/machinery/power/NuclearReactor/proc/stop()
+	STOP_PROCESSING(SSmachines,src)
+	Heat = 0
+	HeatRate = 0
+
+//obj/machinery/power/NuclearReactor/proc/GetHeat()
+//	return HeatRate*Time
+
+/obj/machinery/power/NuclearReactor/proc/GetHeatIncreaseCoeff() //Where C is how strong the reaction is
+	return react.GetFuelAmounts() //So let's pretend we have 2 CRs and 6 FRs, 2 CRs subtracts 2 heat from this tick, but the FRs add 10 more, so you get +10 heat per tick. Heat is good for the reactor, but too much = the big boom boom
+
+//As we know from physics class, efficiency is helped by cooling it down, so at this point in the reaction your efficiency will be helped by some nice n2, for now we'll simulate n2 with vars.
+/obj/machinery/power/NuclearReactor/proc/GetEfficiency(var/temp = 1,var/amt = 1) //As a percent
+	return (1/temp) * amt
+
+/datum/nuclearreaction
+	var/name = "nuclear reaction"
+	var/list/obj/item/twohanded/required/ControlRod/CRS = list()
+	var/list/obj/item/twohanded/required/FuelRod/FRS = list()
+	var/obj/machinery/power/NuclearReactor/reactor
+
+/datum/nuclearreaction/proc/GetCoolAmounts()
+	var/AmtSubtract
+	var/frum = 0
+	for(var/obj/item/twohanded/required/ControlRod/CR in CRS)
+		if(!istype(CR,/obj/item/twohanded/required/ControlRod))
+			return
+		AmtSubtract += CR.ControlPower
+	for(var/obj/item/twohanded/required/FuelRod/FR in FRS)
+		if(!istype(FR,/obj/item/twohanded/required/FuelRod))
```

Comment:
and this

Reply:
It is, byond still cocks it up even when given a specific type.